### PR TITLE
[Feature][connector][kafka] Support read debezium format message from kafka

### PR DESCRIPTION
**Pull request by TaoZex**
## Purpose of this pull request
https://github.com/apache/incubator-seatunnel/issues/3743
Support read debezium format message from kafka

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by TaoZex** [comment]
I will add e2e test for this feature.

**Comment by hailin0** [comment]
> I will add e2e test for this feature.

reference
https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-sqlserver-e2e/src/test/java/org/apache/seatunnel/e2e/connector/cdc/sqlserver/SqlServerCDCIT.java

**Comment by hailin0** [comment]
> > I will add e2e test for this feature.
> 
> reference https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-sqlserver-e2e/src/test/java/org/apache/seatunnel/e2e/connector/cdc/sqlserver/SqlServerCDCIT.java

https://debezium.io/documentation/reference/2.1/tutorial.html#starting-services

**Comment by 2000liux** [comment]
nice。

**Review by TyrantLucifer — COMMENTED** [review]
Could you please add e2e test case about this pull request?

**Comment by hailin0** [comment]
> Could you please add e2e test case about this pull request?

@TaoZex 
reference check cdc sink output

https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcSinkCDCChangelogIT.java#L87

**Comment by hailin0** [comment]
reference this pr start mysql docker & kafka docker & debezium docker

https://github.com/apache/incubator-seatunnel/pull/3950
https://debezium.io/documentation/reference/2.1/tutorial.html#starting-services

**Comment by hailin0** [comment]
add docs into docs/en/connector-v2/formats/debezium-json.md and link to kafka source docs


reference
https://nightlies.apache.org/flink/flink-docs-release-1.16/docs/connectors/table/formats/debezium/

**Comment by TaoZex** [comment]
> reference this pr start mysql docker & kafka docker & debezium docker
> 
> #3950 https://debezium.io/documentation/reference/2.1/tutorial.html#starting-services

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/DebeziumToKafkaIT.java [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+                        "{\"data\":{\"id\":108,\"name\":\"jacket\",\"description\":\"water resistent black wind breaker\",\"weight\":\"0.1\"},\"type\":\"INSERT\"}",
+                        "{\"data\":{\"id\":109,\"name\":\"spare tire\",\"description\":\"24 inch spare tire\",\"weight\":\"22.2\"},\"type\":\"INSERT\"}",
+                        "{\"data\":{\"id\":101,\"name\":\"scooter\",\"description\":\"Small 2-wheel scooter\",\"weight\":\"3.14\"},\"type\":\"DELETE\"}",
+                        "{\"data\":{\"id\":101,\"name\":\"scooter\",\"description\":\"Small 2-wheel scooter\",\"weight\":\"4.56\"},\"type\":\"INSERT\"}",
+                        "{\"data\":{\"id\":107,\"name\":\"rocks\",\"description\":\"box of assorted rocks\",\"weight\":\"5.3\"},\"type\":\"DELETE\"}",
+                        "{\"data\":{\"id\":107,\"name\":\"rocks\",\"description\":\"box of assorted rocks\",\"weight\":\"7.88\"},\"type\":\"INSERT\"}",
+                        "{\"data\":{\"id\":109,\"name\":\"spare tire\",\"description\":\"24 inch spare tire\",\"weight\":\"22.2\"},\"type\":\"DELETE\"}");
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/DebeziumToKafkaIT.java]: test update action ?

**Review by hailin0 — COMMENTED** [review]
add docs & update `release-note.md`

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonFormatOptions.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                    .defaultValue(false)
+                    .withDescription(
+                            "When setting up a Debezium Kafka Connect, users can enable "
+                                    + "a Kafka configuration 'value.converter.schemas.enable' to include schema in the message. "
+                                    + "This option indicates the Debezium JSON data include the schema in the message or not. "
+                                    + "Default is false.");
+
+    public static boolean getSchemaInclude(Map<String, String> options) {
+        return Boolean.parseBoolean(
+                options.getOrDefault(SCHEMA_INCLUDE.key(), SCHEMA_INCLUDE.toString()));
```
Comment (TyrantLucifer) [inline seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonFormatOptions.java]: SCHEMA_INCLUDE.defaultValue() ?

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonFormatOptions.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    public static boolean getIgnoreParseErrors(Map<String, String> options) {
+        return Boolean.parseBoolean(
+                options.getOrDefault(IGNORE_PARSE_ERRORS.key(), IGNORE_PARSE_ERRORS.toString()));
```
Comment (TyrantLucifer) [inline seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonFormatOptions.java]: IGNORE_PARSE_ERRORS.defaultValue()?

File: seatunnel-formats/seatunnel-format-json/src/test/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonSerDeSchemaTest.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                        "{\"before\":{\"id\":111,\"name\":\"scooter\",\"description\":\"Big 2-wheel scooter \",\"weight\":5.17},\"after\":null,\"op\":\"d\"}");
+        assertEquals(expected, actual);
+    }
+    // --------------------------------------------------------------------------------------------
+    // Utilities
+    // --------------------------------------------------------------------------------------------
+
+    private static List<String> readLines(String resource) throws IOException {
+        final URL url = DebeziumJsonSerDeSchemaTest.class.getClassLoader().getResource(resource);
+        assert url != null;
```
Comment (TyrantLucifer) [inline seatunnel-formats/seatunnel-format-json/src/test/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonSerDeSchemaTest.java]: Assertsions.nonNull

Comment (TaoZex) [reply seatunnel-formats/seatunnel-format-json/src/test/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonSerDeSchemaTest.java]: Done. Thanks for your advice.

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonDeserializationSchema.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    @Override
+    public void deserialize(byte[] message, Collector<SeaTunnelRow> out) throws IOException {
+        if (message == null || message.length == 0) {
+            // skip tombstone messages
+            return;
+        }
+
+        try {
+            SeaTunnelRow row = jsonDeserializer.deserialize(message);
```
Comment (hailin0) [inline seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonDeserializationSchema.java]: reference

https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/canal/CanalJsonDeserializationSchema.java#L116
https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/CanalToKafkaIT.java#L265

**Comment by hailin0** [comment]
check ci error

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonDeserializationSchema.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    @Override
+    public SeaTunnelRow deserialize(byte[] message) throws IOException {
+        throw new UnsupportedOperationException(
+                "Please invoke DeserializationSchema#deserialize(byte[], Collector<SeaTunnelRow>) instead.");
+    }
+
+    @Override
+    public void deserialize(byte[] message, Collector<SeaTunnelRow> out) throws IOException {
```
Comment (hailin0) [inline seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonDeserializationSchema.java:68]: example
```java
public void deserialize(byte[] message, Collector<SeaTunnelRow> out) throws IOException {
    if (message == null || message.length == 0) {
            // skip tombstone messages
            return;
    }

    ObjectNode jsonNode = (ObjectNode) convertBytes(message);
    String op = jsonNode.get("op").asText();
    switch (op) {
        case "r":
        case "c":
            SeaTunnelRow insert = convertJsonNode(jsonNode.get("after"));
            insert.setRowKind(RowKind.INSERT);
            out.collect(insert);
            break;
        case "u":
            SeaTunnelRow before = convertJsonNode(jsonNode.get("before"));;
            before.setRowKind(RowKind.UPDATE_BEFORE);
            out.collect(before);

            SeaTunnelRow after = convertJsonNode(jsonNode.get("after"));;
            after.setRowKind(RowKind.UPDATE_AFTER);
            out.collect(after);
            break;
        case "d":
            SeaTunnelRow delete = convertJsonNode(jsonNode.get("after"));;
            before.setRowKind(RowKind.DELETE);
            out.collect(after);
            break;
        }
}
```

Comment (hailin0) [reply seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonDeserializationSchema.java:68]: @TaoZex

Comment (TaoZex) [reply seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonDeserializationSchema.java:68]: Thanks for your advice.

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonSerializationSchema.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import org.apache.seatunnel.api.table.type.SeaTunnelRow;
+import org.apache.seatunnel.api.table.type.SeaTunnelRowType;
+import org.apache.seatunnel.common.exception.CommonErrorCode;
+import org.apache.seatunnel.format.json.JsonSerializationSchema;
+import org.apache.seatunnel.format.json.exception.SeaTunnelJsonFormatException;
+
+import static org.apache.seatunnel.api.table.type.BasicType.STRING_TYPE;
+import static org.apache.seatunnel.format.json.debezium.DebeziumJsonFormatOptions.GENERATE_ROW_SIZE;
+
+public class DebeziumJsonSerializationSchema implements SerializationSchema {
```
Comment (hailin0) [inline seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/debezium/DebeziumJsonSerializationSchema.java:31]: remove this class

**Comment by hailin0** [comment]
please rebase dev branch

**Comment by TaoZex** [comment]
Wait for https://github.com/apache/incubator-seatunnel/pull/4524 merge.

**Comment by EricJoy2048** [comment]
Can you rebase dev branch and push again?

**Review by hailin0 — COMMENTED** [review]
check ci

https://github.com/apache/incubator-seatunnel/actions/runs/4959563825/jobs/8936251986?pr=3981

**Comment by hailin0** [comment]
image

**Comment by hailin0** [comment]
image

**Comment by zhilinli123** [comment]
check CI @TaoZex

**Comment by hailin0** [comment]
Sorry, this PR has been merged

https://github.com/apache/seatunnel/pull/5066

**Pull request closed without merging**